### PR TITLE
[KAT-4104] Add edge counting methods

### PR DESCRIPTION
**Pull request by aneeshdurg**
Previously to count the number of edge AttributedGraph was materializing
a list of all edges and finding the count of that list. This method will
allow just getting a count without retrieving edge ids.

JIRA: KAT-4104

### CODE REVIEW

**Review by rob-isaac — COMMENTED** [review]
Please rename to CountOutEdges

**Comment by rob-isaac** [comment]
> Please rename to CountOutEdges

Sorry I skimmed this too fast and didn't realize it was between two nodes. `CountEdges` is better. Looks like you have a typo `CountOuEdges`

File: libgraph/include/katana/GraphTopology.h [hunk 1]
```diff
@@ -773,6 +773,29 @@ class KATANA_EXPORT EdgeTypeAwareTopology : public EdgeShuffleTopology {
...
+  /// @returns the number of edges from src to dst with edge_type
+  size_t CountEdges(
+      Node src, Node dst, const EntityTypeID& edge_type) const {
+    auto e_range = OutEdges(src, edge_type);
+    if (e_range.empty()) {
+      return 0;
+    }
+
+    internal::EdgeDestComparator<EdgeTypeAwareTopology> comp{this};
+    auto lower_it = std::lower_bound(e_range.begin(), e_range.end(), dst, comp);
```
Comment (amberhassaan) [inline libgraph/include/katana/GraphTopology.h]: Wouldn't it be better to use `std::equal_range`?

File: libgraph/include/katana/GraphTopology.h [hunk 2]
```diff
@@ -773,6 +773,29 @@ class KATANA_EXPORT EdgeTypeAwareTopology : public EdgeShuffleTopology {
...
+    if (lower_it == e_range.end() || OutEdgeDst(*lower_it) != dst) {
+      return 0;
+    }
+
+    auto upper_it = std::upper_bound(lower_it, e_range.end(), dst, comp);
+    return upper_it - lower_it + 1;
```
Comment (amberhassaan) [inline libgraph/include/katana/GraphTopology.h]: why do need to add 1. Why not simply use `std::distance()`?

File: libgraph/include/katana/GraphTopology.h [hunk 3]
```diff
@@ -1223,9 +1246,29 @@ class KATANA_EXPORT EdgeTypeAwareBiDirTopology
...
+  ///
+  /// @param src source node
+  /// @param dst destination node
+  /// @param edge_type edge_type of the edge
+  /// @returns the number of edges from src to dst with edge_type
+  size_t CountEdges(Node src, Node dst, const EntityTypeID& edge_type) const {
+    const auto d_out = OutDegree(src, edge_type);
+    const auto d_in = InDegree(dst, edge_type);
+    if (d_out == 0 || d_in == 0) {
+      return false;
```
Comment (amberhassaan) [inline libgraph/include/katana/GraphTopology.h]: return 0.

**Review by amberhassaan — APPROVED** [review]
Looks good. Left some comments and questions.

**Pull request merged**